Qdrant Semantic Caching

haadirakhangi opened the pull request:
## Title

Semantic Caching with Qdrant Vector database

## Relevant issues

Fixes #4963

## Type

🆕 New Feature
📖 Documentation
✅ Test

## Changes

litellm/caching.py
litellm/utils.py
docs/my-website/docs/caching/all_caches.md

## Testing - Screenshot of new tests passing locally

![image](https://github.com/user-attachments/assets/b026c7cd-d3d2-46a8-8e0c-50fc67e4fcd1)
![image](https://github.com/user-attachments/assets/951fd22f-2d78-491f-af12-5a06805c4b63)

haadirakhangi committed:
qdrant semantic caching added

haadirakhangi committed:
Merge branch 'BerriAI:main' into main

krrishdholakia commented:
picking this up this week - hoping to have this / some version of this merged by end of week.

haadirakhangi commented:
Sure! Let me know if there is anything.

ishaan-jaff reviewed (changes requested):
- please add testing for this PR @haadirakhangi 
- is it possible to add Qdran without adding `qdrant_client` as a dep. We'd prefer to not add new requirements to our Dockerfile

haadirakhangi commented:
Thank you for the feedback, @ishaan-jaff. I'll go ahead and add the necessary tests to the PR.

Regarding the inclusion of `qdrant_client` as a dependency, we can interact with Qdrant via its REST API (please refer to the link below). We can utilize the `requests` library in Python for this purpose. Let me know if this approach works for you, and I'll update the PR accordingly.

https://api.qdrant.tech/api-reference

ishaan-jaff commented:
@haadirakhangi yes please use their REST API 

- don't use requests 
- use our httpx handler litellm has a handler as part of the lib use 
```
client = (
                    _get_async_httpx_client()
                )  # Create a new client if none provided
```
example 
https://github.com/BerriAI/litellm/blob/b1bed459b4efc75f07a7d06b5260deb1a8ce187b/litellm/llms/sagemaker.py#L466

haadirakhangi commented:
Thank you, @ishaan-jaff  I'll proceed with using the REST API and the httpx handler from litellm.

haadirakhangi committed:
implemented RestAPI and added support for cloud and local Qdrant clusters

haadirakhangi committed:
added put method in http_handler

haadirakhangi commented:
Hi @ishaan-jaff,

I've implemented Qdrant Semantic Caching using their REST API and added the 'PUT' method to `http_handler.py`, as it was required for certain actions with Qdrant. You can refer to their API documentation here: [Qdrant API Reference](https://api.qdrant.tech/api-reference).

I’ve tested the implementation on both Qdrant Cloud cluster instances and local cluster instances, and everything is functioning as expected. Could you please guide me on how to properly add tests for this PR?

ishaan-jaff commented:
hi @haadirakhangi looks awesome thanks for the work on this

ishaan-jaff commented:
> Could you please guide me on how to properly add tests for this PR?

what guide are you looking for? Check out test_caching.py -> that should be a good starting point 

- just add testing for `litellm.acompletion` + `litellm.acompletion with streaming` for this PR @haadirakhangi

haadirakhangi committed:
added testing for qdrant semantic caching

ishaan-jaff commented:
can you share a screenshot of both tests working for you locally ?

haadirakhangi commented:
@ishaan-jaff Added the two mentioned tests in the `test_caching.py` file. Please go through it and let me know if there are any changes!

ishaan-jaff commented on docs/my-website/docs/caching/all_caches.md:
```diff
@@ -144,7 +144,62 @@ assert response1.id == response2.id
+random_number = random.randint(1, 100000)
+
+response2 = completion(
+    model="gpt-3.5-turbo",
```
can you pass `mock_response=hi` we don't want the caching tests to make a call to the LLM API

ishaan-jaff replied:
mock_response is a param you can pass to litellm.completion

haadirakhangi replied:
Apparently when I provide two different mock responses (let's say 'hi1' and 'hi2') to the two acompletion request in the first test, it raises an assertion error as both the requests return 'hi1' and 'hi2' respectively. However, if I provide a mock response only to the first acompletion request (only 'hi1'), it is being cached in the second response perfectly. 

Did not face this issue for the second test.

ishaan-jaff reviewed (changes requested):
adjust testing please

haadirakhangi commented:
> can you share a screenshot of both tests working for you locally ?

Sure! Here are the screenshots you asked for:

Testing with acompletion:
![image](https://github.com/user-attachments/assets/d55fac05-0efe-4e7f-8d45-6a5dbec968ed)

Testing with acompletion + streaming:
![image](https://github.com/user-attachments/assets/09a232f7-86a6-40d8-9eda-192c5c1c3f89)

Also, I will adjust the testing as said.

krrishdholakia commented on litellm/caching.py:
```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+        self.similarity_threshold = similarity_threshold
+        self.embedding_model = embedding_model
+
+        if host_type=="cloud":
```
why do i need a 'host_type'? 

is setting base url not sufficient?

haadirakhangi replied:
Base URL works as well. Will update this part soon!

krrishdholakia commented on litellm/caching.py:
```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+        messages = kwargs["messages"]
+        prompt = ""
+        for message in messages:
+            prompt += message["content"]
```
this will cause errors when message["content"] is list

I'd recommend using `get_formatted_prompt` from utils.py

https://github.com/BerriAI/litellm/blob/417547b6f991e05ff6f6663b783e7f8cc889ac96/litellm/utils.py#L4421

haadirakhangi replied:
Noted. Will look into it and update accordingly.

krrishdholakia commented on litellm/caching.py:
```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+        # create an embedding for prompt
+        embedding_response = litellm.embedding(
+            model=self.embedding_model,
+            input=prompt,
```
does the model have context window limits? I'm curious if we can allow the user to control which messages are used in the embedding request (e.g. last user message, etc.) 

Would you want that control? @haadirakhangi

haadirakhangi replied:
That would be great to have. However, might have to think about what is being used to create the embeddings. For instance, if we create an embedding request with last assistant response, semantic caching will most likely fail. Adding last user message makes sense though and it was something I was planning to look into later on.

krrishdholakia commented on litellm/caching.py:
```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+        messages = kwargs["messages"]
+        prompt = ""
+        for message in messages:
+            prompt += message["content"]
```
same issue re: message["content"] being a list

krrishdholakia replied:
or none (e.g. for tool calls)

krrishdholakia commented on litellm/caching.py:
```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+        messages = kwargs["messages"]
+        prompt = ""
+        for message in messages:
+            prompt += message["content"]
```
same issue re: message["content"] being a list or none

krrishdholakia commented on litellm/caching.py:
```diff
@@ -1217,6 +1217,410 @@ async def async_get_cache(self, key, **kwargs):
+        )
+        if llm_router is not None and self.embedding_model in router_model_names:
+            user_api_key = kwargs.get("metadata", {}).get("user_api_key", "")
+            embedding_response = await llm_router.aembedding(
```
in some cases (above) i've seen litellm.aembedding being used. why the switch between the router and litellm.aembedding? 

If no strong reason, could we use a consistent approach?

ishaan-jaff commented:
merging this, I will take care of:

1. making sure testing passes on ci/cd
2. the issues @krrishdholakia pointed out on message["content"]
3. Adding docs on using with litellm proxy 
4. removing the need for  `host_type`

ishaan-jaff commented:
cc @haadirakhangi thanks for the amazing work 👑

haadirakhangi commented:
Thanks @ishaan-jaff, I'm glad to contribute! 😄
